### DISCUSSION TITLE
Anyone used Celery with RabbitMQ Streams?

### DISCUSSION THREAD

**Discussion by saranglakare** [celery/celery#7329] [Q&A] [answered]
RabbitMQ introduced [streams](https://blog.rabbitmq.com/posts/2021/07/rabbitmq-streams-overview) last year. They claim streams work with AMQP 0.9 and 1.0 as well as mentioned [here](https://rabbitmq.github.io/rabbitmq-stream-java-client/stable/htmlsingle/#other-way-to-use-streams-in-rabbitmq). That is, theoretically we should be able to create a queue backed by a stream, connect as many workers we need to fan-out to the queue and each worker should get the message delivered - all of this using existing Celery + RabbitMQ infra.

My question is, has anyone tried to use streams with celery yet? If so, please share any info on how to configure streams in Celery and your experience with them so far. I could not find any blog post or documentation on this.

Thanks in advance!

**Comment by saranglakare** [comment]
Update: I have tried to create a Queue backed by a `stream` using
`Queue('stream-test', routing_key='stream_test', durable=True, queue_arguments={'x-queue-type': 'stream'})`

When I tried to run `celery worker -Q stream-test` I get the following error:

`PreconditionFailed: Basic.consume: (406) PRECONDITION_FAILED - consumer prefetch count is not set for 'queue 'stream-test' in vhost '/''`

Any ideas what I am doing wrong? Seems so close, yet so far!

**Reply by redthing1** [reply]
I get this too.

**Reply by h0jeZvgoxFepBQ2C** [reply]
Did you find a solution? I also have this now?

**Reply by saranglakare** [reply]
> Did you find a solution? I also have this now?

Nope :(

**Comment by saranglakare** [comment, marked as answer]
Update 2: I finally got celery to work with RabbitMQ stream!

The problem is streams do not support [Global QoS](https://www.rabbitmq.com/streams.html#global-qos). Hence, we need to disable it and there's no clean way to do it. Here's a hack posted in https://github.com/celery/celery/issues/6067#issuecomment-1367708093

```python
from celery import bootsteps

class NoChannelGlobalQoS(bootsteps.StartStopStep):
    requires = {'celery.worker.consumer.tasks:Tasks'}

    def start(self, c):
        qos_global = False

        c.connection.default_channel.basic_qos(
            0, c.initial_prefetch_count, qos_global,
        )

        def set_prefetch_count(prefetch_count):
            return c.task_consumer.qos(
                prefetch_count=prefetch_count,
                apply_global=qos_global,
            )

app.steps['consumer'].add(NoChannelGlobalQoS)
```

By putting the above code in `celery.py` of my app (where celery app is initialized) I was able to get it to work.

My queue is declared as:

```python
Queue('streamtest', routing_key='streamtest', durable=True, queue_arguments={'x-queue-type': 'stream'})
```

and now I can run worker to start consuming from this stream:
```
celery -A app_name worker -Q streamtest -c 1
```
(replace `app_name` with your app's name)

The author of the hack mentions that this code creates issues with ETA tasks. So I guess this hack should only be used when starting a worker that is only consuming tasks from a stream and nothing else.

I will now do more experimentation with how tasks work with streams.
